lib/builder/syncer.go

```diff
@@ -102,3 +159,66 @@ func (s *syncer) Sync(builder *Builder) error {
 		Parallel:    builder.VendorReq.Parallel,
 	}, builder.Manifest)
 }
+
+// packSyncer synchronizes local package cache with pack/apps services
+type packSyncer struct {
+	pack pack.PackageService
+	apps app.Applications
+	repo string
+}
+
+// NewPackSyncer creates a new syncer from provided pack and apps services
+func NewPackSyncer(pack pack.PackageService, apps app.Applications, repo string) *packSyncer {
+	return &packSyncer{
+		pack: pack,
+		apps: apps,
+		repo: repo,
+	}
+}
+
+// GetRepository returns the syncer's repository address
+func (s *packSyncer) GetRepository() string {
+	return s.repo
+}
+
+// SelectRuntime picks an appropriate runtime for the application that's
+// being built
+func (s *packSyncer) SelectRuntime(builder *Builder) (*semver.Version, error) {
+	// determine version of this binary
+	teleVersion, err := semver.NewVersion(version.Get().Version)
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to determine tele version")
+	}
+	// determine the latest runtime compatible with this tele
+	runtime, err := pack.FindLatestCompatiblePackage(s.pack, loc.Runtime, *teleVersion)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return runtime.SemVer()
+}
+
+// Sync makes sure that local cache has all required dependencies for the
+// selected runtime
+func (s *packSyncer) Sync(builder *Builder, runtimeVersion *semver.Version) error {
+	cacheApps, err := builder.Env.AppServiceLocal(localenv.AppConfig{})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	err = service.PullAppDeps(service.AppPullRequest{
+		SrcPack:     s.pack,
+		SrcApp:      s.apps,
+		DstPack:     builder.Env.Packages,
+		DstApp:      cacheApps,
+		Parallel:    builder.VendorReq.Parallel,
+		FieldLogger: builder.FieldLogger,
+	}, builder.Manifest)
+	if err != nil {
+		if utils.IsNetworkError(err) || trace.IsEOF(err) {
+			return trace.ConnectionProblem(err, "failed to download "+
+				"application dependencies from %v - please make sure the "+
+				"repository is reachable: %v", builder.Repository, err)
```

Comment:
Is `builder.Repository` meaningful in this error message?

Reply:
Yes, because the same packSyncer is used by enterprise edition to sync with get.gravitational.io.